content/docs/tasks/policy-enforcement/control-headers/index.md

```diff
@@ -0,0 +1,242 @@
+        "Accept": "*/*",
+        "Content-Length": "0",
+        ...
+        "X-Envoy-Internal": "true"
+      }
+    }
+    {{< /text >}}
+
+   The output should be the request headers as they are received by _httpbin_ service.
+
+1. Create a rule for the demo adapter:
+
+    {{< text yaml >}}
+    $ kubectl apply -f - <<EOF
+    apiVersion: config.istio.io/v1alpha2
+    kind: rule
+    metadata:
+      name: keyval
+      namespace: istio-system
+    spec:
+      actions:
+      - handler: keyval.istio-system
+        instances: [ keyval ]
+        name: x
+      requestHeaderOperations:
+      - name: user-group
+        values: [ x.output.value ]
+    EOF
+    {{< /text >}}
+
+1. Issue a new request to the ingress gateway with the header `key` set to value `jason`:
+
+    {{< text bash >}}
+    $ curl -Huser:jason http://$INGRESS_HOST:$INGRESS_PORT/headers
+    {
+      "headers": {
+        "Accept": "*/*",
+        "Content-Length": "0",
+        "User": "jason",
+        "User-Agent": "curl/7.58.0",
+        "User-Group": "admin",
+        ...
+        "X-Envoy-Internal": "true"
+      }
+    }
+    {{< /text >}}
+
+   Note the presence of `user-group` header with the value derived from the
+   rule application of the adapter. The expression `x.output.value` in the rule
+   evaluates to the populated `value` field returned by `keyval` adapter.
+
+1. Modify the rule to rewrite the URI path to a different virtual service route
+   if the check succeeds:
+
+    {{< text yaml >}}
+    $ kubectl apply -f - <<EOF
+    apiVersion: config.istio.io/v1alpha2
+    kind: rule
+    metadata:
+      name: keyval
+      namespace: istio-system
+    spec:
+      match: source.labels["istio"] == "ingressgateway"
+      actions:
+      - handler: keyval.istio-system
+        instances: [ keyval ]
+      requestHeaderOperations:
+      - name: :path
+        values: [ '"/status/418"' ]
+    EOF
+    {{< /text >}}
+
+1. Repeat the request to the ingress gateway:
+
+    {{< text bash >}}
+    $ curl -Huser:jason -I http://$INGRESS_HOST:$INGRESS_PORT/headers
+    HTTP/1.1 418 Unknown
+    server: istio-envoy
+    ...
+    {{< /text >}}
+
+   Note that the ingress gateway changed the route _after_ the rule application
+   of the policy adapter. The modified request may use a different route and
+   destination and is subject to the traffic management configuration.
+
+   The modified request is not checked again by the policy engine within the
+   same proxy. Therefore, we recommend to use this feature in gateways, so
+   that the server-side policy checks take effect.
+
+## Custom error response
+
+The adapter developers can also specify the client-facing error response if the policy
+check fails. Issue a request to the gateway with a non-existing user key:
+
+{{< text bash >}}
+$ curl -Huser:bill http://$INGRESS_HOST:$INGRESS_PORT/headers
+<error_detail>key "bill" not found</error_detail>
+{{< /text >}}
+
+The demo adapter uses a [custom error
```

Comment:
Remove this paragraph and the sample Go code, this is not relevant to customers.

Reply:
Done